write-fonts/src/graph/splitting.rs

```diff
@@ -0,0 +1,402 @@
+//! splitting layout subtables
+
+use std::collections::{HashMap, HashSet};
+
+use font_types::{GlyphId, Offset16};
+
+use super::{Graph, ObjectId};
+use crate::tables::layout as wlayout;
+use crate::types::FixedSize;
+use crate::write::TableData;
+
+use crate::read::tables::gpos as rgpos;
+use crate::read::tables::layout as rlayout;
+
+const MAX_TABLE_SIZE: usize = u16::MAX as usize;
+
+pub(super) fn split_pair_pos(graph: &mut Graph, lookup: ObjectId) {
+    split_subtables(graph, lookup, split_pair_pos_subtable)
+}
+
+/// A common impl handling updating the lookuup with the new subtables
+fn split_subtables(
+    graph: &mut Graph,
+    lookup: ObjectId,
+    split_fn: fn(&mut Graph, ObjectId) -> Option<Vec<ObjectId>>,
+) {
+    let data = graph.objects.remove(&lookup).unwrap();
+    debug_assert!(data.reparse::<rgpos::PositionLookup>().is_ok());
+    log::debug!("trying to split subtables in {lookup:?}, '{}'", data.type_);
+
+    let mut new_subtables = HashMap::new();
+    for subtable in &data.offsets {
+        if let Some(split_subtables) = split_fn(graph, subtable.object) {
+            new_subtables.insert(subtable.object, split_subtables);
+        }
+    }
+
+    if new_subtables.is_empty() {
+        // just put the old data back unchanged; nothing to see here
+        graph.objects.insert(lookup, data);
+        log::debug!("nothing to split, continuing");
+        return;
+    }
+
+    let n_new_subtables = new_subtables
+        .values()
+        // - 1 because each group of new subtables replaces an old subtable
+        .map(|ids| ids.len() - 1)
+        .sum::<usize>();
+
+    let n_total_subtables: u16 = (data.offsets.len() + n_new_subtables).try_into().unwrap();
+    // we just want the lookup type/flag/etc, but we need a generic FontRead type
+    let generic_lookup: rlayout::Lookup<()> = data.reparse().unwrap();
+    let mut new_data = TableData::new(data.type_);
+    new_data.write(generic_lookup.lookup_type());
+    new_data.write(generic_lookup.lookup_flag());
+    new_data.write(n_total_subtables);
+    for sub in data.offsets {
+        match new_subtables.get(&sub.object) {
+            Some(new) => new.iter().for_each(|id| new_data.add_offset(*id, 2, 0)),
+            None => new_data.add_offset(sub.object, 2, 0),
+        }
+    }
+
+    graph.nodes.get_mut(&lookup).unwrap().size = new_data.bytes.len() as _;
+    graph.objects.insert(lookup, new_data);
+}
+
+fn split_pair_pos_subtable(graph: &mut Graph, lookup: ObjectId) -> Option<Vec<ObjectId>> {
+    let data = &graph.objects[&lookup];
+    let format: u16 = data.read_at(0).unwrap();
+    match format {
+        1 => split_pair_pos_format_1(graph, lookup),
+        2 => {
+            log::info!("table splitting not yet implemented for PairPos format 2");
+            None
+        }
+        _ => unreachable!("we only write valid formats"),
```

Comment:
Are we certain that this won’t be encountered when subsetting an arbitrary font? We don’t want to panic in that case. 

Maybe we can filter out subtables with invalid formats and emit a warning.